src/codemodder/dependency_management/pyproject_writer.py

```diff
@@ -0,0 +1,55 @@
+import tomlkit
+from typing import Optional
+from copy import deepcopy
+from codemodder.dependency import Dependency
+from codemodder.change import Action, Change, ChangeSet, PackageAction, Result
+from codemodder.dependency_management.base_dependency_writer import DependencyWriter
+from codemodder.diff import create_diff_and_linenums
+
+
+class PyprojectWriter(DependencyWriter):
+    def add_to_file(
+        self, dependencies: list[Dependency], dry_run: bool = False
+    ) -> Optional[ChangeSet]:
+        pyproject = self._parse_file()
+        original = deepcopy(pyproject)
+
+        try:
+            pyproject["project"]["dependencies"].extend(
+                [f"{dep.requirement}" for dep in dependencies]
+            )
+        except tomlkit.exceptions.NonExistentKey:
+            return None
+
+        diff, added_line_nums = create_diff_and_linenums(
+            tomlkit.dumps(original).split("\n"), tomlkit.dumps(pyproject).split("\n")
+        )
+
+        if not dry_run:
+            with open(self.path, "w", encoding="utf-8") as f:
+                tomlkit.dump(pyproject, f)
+
+        changes = [
+            Change(
```

Comment:
I think this should be factored out into the base class now since it includes some fiddly logic that will be pretty much the same for each writer.